# SQL Murder Mystery Practice (Solving)

---

## Problem Description

A crime has taken place and the detective needs your help. The detective gave you the crime scene report, but you somehow lost it. You vaguely remember that the crime was a ​murder​ that occurred sometime on **Jan.15, 2018** and that it took place in **SQL City**. Start by retrieving the corresponding crime scene report from the police department’s database.

---

## Import Necessary Libraries

In [1]:
# For Data Frame
import pandas as pd

# For The Database Connection
import mysql.connector as my_conn

# For Table Formatted Data Frame Printing
from rich.console import Console
from rich.table import Table


## Connect The MySQL Database

In [2]:
# connecting MySQL Database To Python
db = my_conn.connect(
    host = "localhost",
    user = "root",
    password = "root",
    database = "sql_city",
    consume_results=True
)

In [3]:
db_cursor = db.cursor() 

## Table Content Exploration In The sql_city Database 

In [4]:
# Getting table names in the sql_city database in the Local MySQL Database Management System
i = 1
db_cursor.execute("SHOW TABLES")
for x in db_cursor:
    print(i, ": ", x[0])
    i+=1

1 :  crime_scene_report
2 :  drivers_license
3 :  facebook_event_checkin
4 :  get_fit_now_check_in
5 :  get_fit_now_member
6 :  income
7 :  interview
8 :  person
9 :  solutions


---

In [5]:
# Getting data from each row for deeper investigation
db_cursor.execute("SHOW TABLES")
tables = db_cursor.fetchall()

for i in tables:
    db_cursor.execute("SELECT * FROM %s "  %i[0])
    data = db_cursor.fetchall()
    columns = db_cursor.column_names
    
    console = Console()
    table = Table(i[0])
    df = pd.DataFrame(data=data, columns=columns).head()
    table.add_row(df.to_string())
    console.print(table)

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ crime_scene_report                                                                      ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        date     type                                              description      city │
│ 0  20180115  robbery        A Man Dressed as Spider-Man Is on a Robbery Spree       NYC │
│ 1  20180115   murder                        Life? Dont talk to me about life.    Albany │
│ 2  20180115   murder      Mama, I killed a man, put a gun against his head...      Reno │
│ 3  20180215   murder                               REDACTED REDACTED REDACTED  SQL City │
│ 4  20180215   murder  Someone killed the guard! He took an arrow to the knee!  SQL City │
└─────────────────────────────────────────────────────────────────────────────────────────┘

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ drivers_license                                                                      ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        id  age  height eye_color hair_color  gender plate_number  car_make car_model │
│ 0  100280   72      57     brown        red    male       P24L4U     Acura       MDX │
│ 1  100460   63      72     brown      brown  female       XF02T6  Cadillac       SRX │
│ 2  101029   62      74     green      green  female       VKY5KR     Scion        xB │
│ 3  101198   43      54     amber      brown  female       Y5NZ08    Nissan     Rogue │
│ 4  101255   18      79      blue       grey  female       5162Z1     Lexus        GS │
└──────────────────────────────────────────────────────────────────────────────────────┘

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ facebook_event_checkin                                                                                ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│    person_id  event_id                                                           event_name      date │
│ 0      28508      5880                      Nudists are people who wear one-button suits.\n  20170913 │
│ 1      63713      3865  but that's because it's the best book on anything for the layman.\n  20171009 │
│ 2      63713      3999                             If Murphy's Law can go wrong, it will.\n  20170502 │
│ 3      63713      6436     Old programmers never die.  They just branch to a new address.\n  20170926 │
│ 4      82998      4470                                 Help a swallow land at Capistrano.\n  20171022 │
└───────────────────────────────────────────────────────────────────────────────────────────────────────┘

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ get_fit_now_check_in                                          ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│   membership_id  check_in_date  check_in_time  check_out_time │
│ 0         NL318       20180212            329             365 │
│ 1         NL318       20170811            469             920 │
│ 2         NL318       20180429            506             554 │
│ 3         NL318       20180128            124             759 │
│ 4         NL318       20171027            418            1019 │
└───────────────────────────────────────────────────────────────┘

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ get_fit_now_member                                                              ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│       id  person_id               name  membership_start_date membership_status │
│ 0  03487      73490  Lyman Harshbarger               20170529            silver │
│ 1  06V71      50993      Travis Windly               20170812              gold │
│ 2  07QP3      15980       Hugh Piascik               20171222           regular │
│ 3  0862D      62171       Johana Dayer               20170612              gold │
│ 4  0CR33      62561         Noe Elgart               20170627              gold │
└─────────────────────────────────────────────────────────────────────────────────┘

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ income                      ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│          ssn  annual_income │
│ 0  100009868          52200 │
│ 1  100169584          64500 │
│ 2  100300433          74400 │
│ 3  100355733          35900 │
│ 4  100366269          73000 │
└─────────────────────────────┘

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ interview                                                                             ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│    person_id                                                               transcript │
│ 0      28508                                      ‘I deny it!’ said the March Hare.\n │
│ 1      63713                                                                       \n │
│ 2      86208                            way, and the whole party swam to the shore.\n │
│ 3      35267  lessons in here? Why, there’s hardly room for YOU, and no room at all\n │
│ 4      33856                                                                       \n │
└───────────────────────────────────────────────────────────────────────────────────────┘

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ person                                                                                   ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│       id                 name  license_id  address_number address_street_name        ssn │
│ 0  10000   Christoper Peteuil      993845             624        Bankhall Ave  747714076 │
│ 1  10007  Kourtney Calderwood      861794            2791       Gustavus Blvd  477972044 │
│ 2  10010            Muoi Cary      385336             741     Northwestern Dr  828638512 │
│ 3  10016          Era Moselle      431897            1987       Wood Glade St  614621061 │
│ 4  10025         Trena Hornby      550890             276       Daws Hill Way  223877684 │
└──────────────────────────────────────────────────────────────────────────────────────────┘

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ solutions                                                                                                       ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│                                user                                                                             │
│ value                                                                                                           │
│ 0        4A6572656D7920426F77657273  Congrats, you found the murderer! But wait, there is more... If you think  │
│ you are up for a challenge, try querying the interview transcript of the murderer to find the real villain      │
│ behind this crime. If you feel especially confident in your SQL skills, try to complete this final step with no │
│ more than 2 queries.                                                                                            │
│ 1  4D6972616E6461205072696573746C79                                                                             │
│ Congrats, you found the brains behind the murder! Everyone in SQL City hails you as the greatest SQL detective  │
│ of all time. Time to break out the champagne!                                                                   │
│ 2                          4E45494E                                                                             │
│ That is not the right person. Try again!                                                                        │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

---

## Steps For Enlightening The Murder Mystery

### Step1: Get Crime Scene Report By Using OnHand Information

As it is indicated in the problem description, the murder has been occurred in the **SQL City** on  **Jan.15, 2018**. Therefore, crime details should be pulled from the crime_scene_report with respect to 'SQL City', 'Jan.15, 2018' and 'murder' criterions

In [6]:
# Crime Scene Report
db_cursor.execute("SELECT description FROM crime_scene_report WHERE city = 'SQL City' AND date='20180115' AND type = 'murder'")
print(db_cursor.fetchall()[0][0])

Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".


---

### Step2: Find The Witnesses

To get deeper understanding about the case, two witnesses and their interviews must be found.

In [7]:
# First Witness
# The first witness lives at the last house on "Northwestern Dr".
db_cursor.execute("SELECT * FROM person WHERE address_street_name='Northwestern Dr' ORDER BY address_number DESC LIMIT 3")
pd.DataFrame(data=db_cursor.fetchall(), columns=db_cursor.column_names)

id            name  license_id  address_number address_street_name  \
0  14887  Morty Schapiro      118009            4919     Northwestern Dr   
1  17729  Lasonya Wildey      439686            3824     Northwestern Dr   
2  53890  Sophie Tiberio      957671            3755     Northwestern Dr   

         ssn  
0  111564949  
1  917817122  
2  442830147

As it is shown in the above data frame, last house on "Northwestern Dr" should have biggest address number. Therefore, it should be **Morty Schapiro** with the **14887 id and 118009 license number**.

In [8]:
# Second Witness
# The second witness, named Annabel, lives somewhere on "Franklin Ave".
db_cursor.execute("SELECT * FROM person WHERE name LIKE '%Annabel%' AND address_street_name='Franklin Ave'")
pd.DataFrame(data=db_cursor.fetchall(), columns=db_cursor.column_names)

id            name  license_id  address_number address_street_name  \
0  16371  Annabel Miller      490173             103        Franklin Ave   

         ssn  
0  318771143

As it is shown in the above data frame, second witness should be **Annabel Miller**. Since, she is the only record in person table who lives somewhere on the 'Franklin Ave' with name similarity.

---

### Step3: Go To Witnesses' Interviews

As second step findings, **First Witness is Morty Schapiro** and **Second Witness is Annabel Miller**.

In [9]:
# First Witness(Morty Schapiro) Interview
db_cursor.execute("SELECT transcript FROM interview WHERE person_id = (SELECT id FROM person WHERE address_street_name='Northwestern Dr' ORDER BY address_number DESC LIMIT 1)")
db_cursor.fetchall()

[('I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".',)]

OnHand Information: **A Male**, "Get Fit Now Gym" bag(**gold member**), membership number with **"48Z"**, car plate with **"H42W"**

In [10]:
# Second Witness(Annabel Miller) Interview
db_cursor.execute("SELECT transcript FROM interview WHERE person_id = (SELECT id FROM person WHERE name LIKE '%Annabel%' AND address_street_name='Franklin Ave')")
db_cursor.fetchall()

[('I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.',)]

OnHand Information: Killer was at the gym on **Jan.09,2018**

---

### Step4: Track Down The Killer From The Gym System And Driver Licence

In [11]:
# Tracking Down The Killer According To Witnesses Interviews
db_cursor.execute("SELECT member.name FROM get_fit_now_member member\
                   INNER JOIN person ON member.person_id=person.id \
                   INNER JOIN drivers_license dl ON dl.id = person.license_id \
                   INNER JOIN get_fit_now_check_in checkin ON checkin.membership_id = member.id \
                   WHERE member.membership_status = 'gold' \
                         AND member.id LIKE '%48Z%' \
                         AND dl.plate_number LIKE '%H42W%' \
                         AND checkin.check_in_date = '20180109'")
killer = db_cursor.fetchall()[0][0]
killer

'Jeremy Bowers'

As a result of witnesses testifications, it is understood that **Jeremy Bowers** is the killer.

---

## Check The Result According To solutions Table In The Database

In [12]:
# Get Answers To A DataFrame From The solutions Table From The Database
db_cursor.execute("SELECT * FROM solutions")
df = pd.DataFrame(data=db_cursor.fetchall(), columns=db_cursor.column_names)
df

user  \
0        4A6572656D7920426F77657273   
1  4D6972616E6461205072696573746C79   
2                          4E45494E   

                                               value  
0  Congrats, you found the murderer! But wait, th...  
1  Congrats, you found the brains behind the murd...  
2           That is not the right person. Try again!

In [13]:
# Check If The Killer That We Found Is The Right Person That Is In The solutions Table
for i in range(0,len(df)):
    if bytes.fromhex(df.loc[i,'user']).decode('utf-8') == killer:
        print(df.loc[i,'value'])
        break
    elif i == 2:
        print(df.loc[i,'value'])

Congrats, you found the murderer! But wait, there is more... If you think you are up for a challenge, try querying the interview transcript of the murderer to find the real villain behind this crime. If you feel especially confident in your SQL skills, try to complete this final step with no more than 2 queries.


### BONUS

It seems the killer is just the tool. Therefore, instigator needs to be found.

In [14]:
# Checking The Interview Of The Killer To Learn Clues About The Instigator
db_cursor.execute("SELECT transcript FROM interview WHERE person_id = (SELECT id FROM person WHERE name='%s')" %killer)
print(db_cursor.fetchall()[0][0])

I was hired by a woman with a lot of money. I don't know her name but I know she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S. I know that she attended the SQL Symphony Concert 3 times in December 2017.



OnHand Information: **Female** / Height  **5'5" (65") or 5'7" (67")** / **Red Hair** / **Tesla Model S** / **SQL Symphony Concert 3 times in December 2017**

In [15]:
# Finding The Instigator By Using Killer Interview Clues
db_cursor.execute("SELECT DISTINCT(person.name) FROM drivers_license dl \
                   INNER JOIN person ON dl.id = person.license_id \
                   INNER JOIN facebook_event_checkin fb ON fb.person_id = person.id \
                   WHERE fb.event_name LIKE '%SQL Symphony Concert%' AND fb.date LIKE '%201712%' AND\
                   dl.gender='female' AND dl.hair_color='red' AND dl.car_make='Tesla' AND dl.car_model = 'Model S' ")
instigator = db_cursor.fetchall()[0][0]
instigator

'Miranda Priestly'

---

In [16]:
# Check If The Instigator That We Found Is The Right Person That Is In The solutions Table
for i in range(0,len(df)):
    if bytes.fromhex(df.loc[i,'user']).decode('utf-8') == instigator:
        print(df.loc[i,'value'])
        break
    elif i == 2:
        print(df.loc[i,'value'])

Congrats, you found the brains behind the murder! Everyone in SQL City hails you as the greatest SQL detective of all time. Time to break out the champagne!


---

## Close The Database Connection

In [17]:
# Close The Database Connection 
db_cursor.close()
db.close()

---